File: v2/90.py






class Solution:
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # 6 star, dfs, 多练习
        nums.sort()
        rs = []
        self.dfs([], rs, 0, nums)
        return rs


    def dfs(self, path, rs, start, nums):
        if len(path) <= len(nums) and path not in rs:
            rs.append(path)
        for i in range(start, len(nums)):
            self.dfs(path+[nums[i]], rs, i+1, nums)

# print(Solution().subsetsWithDup([1,2,2]))

File: go/90.go

//package main

import (
	//"fmt"
	"sort"
)


// 5 star, dfs, 去重比较巧妙
// slow, beat 0%, 8ms
func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	ret := [][]int{}
	retP := &ret
	dfs(retP, nums, []int{}, 0)
	return *retP


}

func dfs(retP *[][]int, nums []int, path []int, start int) {
	temp := make([]int, len(path))
	copy(temp, path)
	*retP = append(*retP, temp)
	if len(nums) == 0 {return}
	for i:= start; i<len(nums); i++ {
		//fmt.Println(start, path)
		if i == start || nums[i] != nums[i-1] {  //注意去除重复
			dfs(retP, nums, append(path, nums[i]), i+1)
		}

	}
}

//func subsetsWithDup(nums []int) [][]int {
//	sort.Ints(nums)
//	res := [][]int{}
//	var dfs func(int, []int)
//	dfs = func(index int, tmp []int) {
//		t := make([]int, len(tmp))
//		copy(t, tmp)
//		res = append(res, t)
//
//		for i := index; i < len(nums); i++ {
//			if i == index || nums[i] != nums[i-1] {
//				dfs(i+1, append(tmp, nums[i]))
//			}
//		}
//	}
//
//	dfs(0, make([]int, 0, len(nums)))
//	return res
//}

//func main() {
//	nums := []int {1,2,2}
//	fmt.Println(subsetsWithDup(nums))
//}
